File: src/lib/api/jumpserver.ts
import { devApiFetch, hasTauriRuntime, invoke } from "./client";
import type {
  JumpServerOpenResult,
  JumpServerSession,
  UpsertJumpServerSessionInput,
} from "@/types";

async function openExternalUrl(url: string) {
  if (hasTauriRuntime()) {
    try {
      const opener = await import("@tauri-apps/plugin-opener");
      await opener.openUrl(url);
      return;
    } catch {
      window.open(url, "_blank", "noopener,noreferrer");
      return;
    }
  }
  window.open(url, "_blank", "noopener,noreferrer");
}

export const jumpserverApi = {
  list: () =>
    hasTauriRuntime()
      ? invoke<JumpServerSession[]>("list_jumpserver_sessions")
      : devApiFetch<JumpServerSession[]>("/jumpserver-sessions"),
  upsert: (input: UpsertJumpServerSessionInput) =>
    hasTauriRuntime()
      ? invoke<JumpServerSession>("upsert_jumpserver_session", { input })
      : devApiFetch<JumpServerSession>("/jumpserver-sessions", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  delete: (key: string) =>
    hasTauriRuntime()
      ? invoke<void>("delete_jumpserver_session", { key })
      : devApiFetch<void>(`/jumpserver-sessions/${encodeURIComponent(key)}`, {
          method: "DELETE",
        }),
  open: async (key: string) => {
    const result = hasTauriRuntime()
      ? await invoke<JumpServerOpenResult>("open_jumpserver_session", { key })
      : await devApiFetch<JumpServerOpenResult>(
          `/jumpserver-sessions/${encodeURIComponent(key)}/open`,
          { method: "POST" },
        );
    await openExternalUrl(result.webUrl);
    return result;
  },
};

File: src-tauri/src/commands/jumpserver.rs
use crate::error::CommandError;
use crate::models::{JumpServerOpenResult, JumpServerSession, UpsertJumpServerSessionInput};
use crate::services::jumpserver::JumpServerService;
use crate::state::AppState;

#[tauri::command]
pub fn list_jumpserver_sessions(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<JumpServerSession>, CommandError> {
    JumpServerService::list(&state.db).map_err(|e| e.into())
}

#[tauri::command]
pub fn upsert_jumpserver_session(
    state: tauri::State<'_, AppState>,
    input: UpsertJumpServerSessionInput,
) -> Result<JumpServerSession, CommandError> {
    JumpServerService::upsert(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn open_jumpserver_session(
    state: tauri::State<'_, AppState>,
    key: String,
) -> Result<JumpServerOpenResult, CommandError> {
    JumpServerService::open(&state.db, &key).map_err(|e| e.into())
}

#[tauri::command]
pub fn delete_jumpserver_session(
    state: tauri::State<'_, AppState>,
    key: String,
) -> Result<(), CommandError> {
    JumpServerService::delete(&state.db, &key).map_err(|e| e.into())
}
